Added support for the auth-api service in the httpd pod.

abellotti opened the pull request:
- Added the auth-api service.
- Updated Dockerfile to:
  - include /opt/rubies 2.3.1 same as ManageIQ/container-ruby
  - bundle install the gems
  - install the auth-api service.

Related PRs:
- [ManageIQ/manageiq-pods PR# 204](https://github.com/ManageIQ/manageiq-pods/pull/204)
- [ManageIQ/container-httpd PR# 15](https://github.com/ManageIQ/container-httpd/pull/15)
- [ManageIQ/manageiq PR# 15881](https://github.com/ManageIQ/manageiq/pull/15881)

abellotti committed:
Added support for the auth-api service in the httpd pod.

- Added the auth-api crystal/kemal source code.
- Updated Dockerfile to build and install the auth-api service.
- Remove the auth-api source and crystal upon completion.

abellotti commented:
WIP - do not merge.

TODO:
- update auth-api service to leverage HTTPD_AUTH_API_SERVICE_PORT from environment
upon starting up the service.
-  testing to be done with latest containers.

carbonin commented:
Why is the source here? Shouldn't we have a separate repo for that?

abellotti committed:
Updated auth-api service so that it leverages the container
environment so its port is driven from the configuration.

abellotti commented:
@carbonin we could, I'm open to suggestions.

one option I have is the following:
- ManageIQ/container-httpd-auth-config  for the Pod used to setup the different ext-auth configurations and generate the auth configmap
- ManageIQ/container-httpd-auth-scripts  for all related scripts/sources/etc. 

both container-httpd-auth-config and container-httpd can git clone the scripts repo during image builds.  we can stash the auth-api crystal source in that scripts repo.

Thoughts ?

carbonin commented:
I'm more concerned about how we install things. Will probably need some input from @simaishi on this one.

bdunne commented on docker-assets/auth-api/src/auth-api/user.cr:
```diff
@@ -0,0 +1,51 @@
+      hash_result = Hash(String, String).new
+      if user_attrs.is_a? Hash(DBus::Type, DBus::Type)
+        user_attrs.each do |key, dbus_value|
+          if dbus_value.is_a?(DBus::Variant)
```
Can these nested `if` conditions be changed to guard clauses to reduce complexity?
```ruby
next unless ...
```

bdunne commented on docker-assets/entrypoint:
```diff
@@ -1,3 +1,5 @@
 #!/bin/bash
 
+env | sort > /etc/container-environment
```
Interesting, does the order matter here?

bdunne commented on Dockerfile:
```diff
@@ -64,7 +64,24 @@ EXPOSE 80
-RUN systemctl enable dbus httpd
+## Build Auth-Api
+COPY docker-assets/auth-api /tmp/auth-api
+RUN (cd /tmp/auth-api                                             && \
```
I assume most of this will go away once we figure out how to build the RPM for crystal, right? (Part of the reason it's marked as WIP)

abellotti replied:
going back to Ruby, WIP was for that decision as well as testing needed with latest containers. I'll need the initial auth-config container and first IPA config CLI to test end to end, then I can verify this PR and unwip at that time. Thanks.

bdunne reviewed (changes requested):
@abellotti Does this need to be in the same container as httpd?  I know it needs dbus and all of the authentication config, but I'm wondering if it should be a separate container in the same pod.  @carbonin What do you think?

abellotti committed:
rewrote auth-api service to be in ruby running with the rh-ruby230 scl ruby.

abellotti commented:
/cc @Fryguy here's the updated service in ruby.

carbonin commented on Dockerfile:
```diff
@@ -73,11 +73,26 @@ COPY docker-assets/initialize-httpd-auth.service /usr/lib/systemd/system/initial
+# Ruby 2.3
+ENV TERM=xterm \
+    LANG=en_US.UTF-8
+RUN yum -y install --setopt=tsflags=nodocs rh-ruby23
```
Instead of using a separate ruby package (scl) can we base this image off of https://github.com/ManageIQ/container-ruby?

abellotti replied:
I thought of that, but can we have 2 FROM lines ?  I really need FROM centos/httpd:latest more.

carbonin replied:
Ah right, then can we at least maintain the same method of installing ruby as https://github.com/ManageIQ/container-ruby/blob/master/Dockerfile?

abellotti replied:
Yep, that's fine.  My only requirement was ruby 2.2 for sinatra, so we should be ok there.
I'll give it a try.

bdunne commented on docker-assets/auth-api/Gemfile:
```diff
@@ -0,0 +1,5 @@
+
```
leading blank line?

Fryguy replied:
LOLvim :trollface:

carbonin replied:
I think something happened here and this is resolved.

Fryguy replied:
yeah, it was a leading blank line, and git re-resolved this somewhere else.

bdunne commented on docker-assets/auth-api/auth_api_service.rb:
```diff
@@ -0,0 +1,55 @@
+  DEFAULT_USER_ATTRIBUTES = %w(mail givenname sn displayname domainname).freeze
+
+  set :bind, "0.0.0.0"
+  set :port, ENV["HTTPD_AUTH_API_SERVICE_PORT"]
```
Does this need to be configurable?

abellotti replied:
Not offhand. unless they really, really want to change it, they can update it in the template and redeploy.

carbonin replied:
I think we should be setting this to `3301` directly. Additionally I don't see it defined in the deployment env section in https://github.com/ManageIQ/manageiq-pods/pull/204 so I'm not sure how this would work as is.

bdunne replied:
Is 3301 the default port for Sinatra?

carbonin replied:
This env variable is provided automatically based on the service presence in the namespace (see: https://docs.openshift.org/latest/dev_guide/environment_variables.html#automatically-added-environment-variables)

abellotti replied:
3301 is not the Sinatra port. It's an unused port as per IANA.

bdunne replied:
Why aren't we using the default port?

abellotti replied:
I didn't want to use the default port,  If another service comes along and uses Sinatra (though unlikely, but you never know), then we'd have to reconfigure.  I'd rather have dedicated ports for the different services we implement.

carbonin replied:
> If another service comes along and uses Sinatra

Another sinatra app running in this container seems exceedingly unlikely.

carbonin replied:
How about 8080

Fryguy commented on docker-assets/auth-api/auth_api_service.rb:
```diff
@@ -0,0 +1,55 @@
+
+    begin
+      ifp_interface = dbus_ifp_interface
+      user_attrs = ifp_interface.GetUserAttr(params[:userid], attrs_needed).first
```
Not sure how this works under the covers, but please be sure there are no injection security concerns here...both values come from params which are user-manipulable.

abellotti replied:
All seemed ok, I've tried with funky characters, ye-old redirects for shell, the & and such for URI, got either blocked early (badly formed URI), sinatra complaining on parameter or rejected from the infopipe library call. 

Regardless, I will ask the dbus/infopipe interface library devs.  This is specifically for org.freedesktop.sssd.infopipe.GetUserAttr and .GetUserGroups.

Fryguy commented on docker-assets/auth-api/auth_api_service.rb:
```diff
@@ -0,0 +1,55 @@
+  get "/api/user_groups/:userid", :provides => 'json' do
+    begin
+      ifp_interface = dbus_ifp_interface
+      user_groups = ifp_interface.GetUserGroups(params[:userid])
```
Same here.

abellotti committed:
- Standardizing on the 2.3.1 /opt/rubies ruby to be consistent with ManageIQ/container-ruby
- fixed up Gemfile

abellotti commented:
/cc @carbonin @Fryguy 
updated to build ruby 2.3.1 and moved MiQ code to lib/httpd_auth_api.rb,
we're still good, tested end to end.

![screen shot 2017-09-09 at 5 06 32 pm](https://user-images.githubusercontent.com/5797328/30243937-f23c30dc-9581-11e7-9c88-f1e34d60462c.png)

abellotti committed:
Moved some of the docker-assets copying and service definitions after the
ruby build and bundle install. Was getting ruby builds with each small
change for services I was making in the docker-assets directory.

carbonin commented on docker-assets/auth-api-service.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/bash
+
+export PATH=$PATH:/opt/rubies/ruby-2.3.1/bin
+cd /opt/rh/auth-api
```
Don't think this should be `rh` probably `manageiq`

bdunne replied:
Or just `/opt/auth-api` since it has nothing to do with `rh` or `manageiq`?

Fryguy replied:
It might be weird to have a top-level "project" called auth-api though?  it's part of manageiq, so I'd prefer manageiq if we namespace at all.  if red hat happens to use it, it's all under the covers anyway.

carbonin commented on docker-assets/auth-api-service.sh:
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/bash
+
+export PATH=$PATH:/opt/rubies/ruby-2.3.1/bin
```
Is this not something that is handled by the `container-environment` file?

It's defined in the dockerfile so I figure it should get dumped into that file.

carbonin commented:
@abellotti Any update on this?

Fryguy commented on Dockerfile:
```diff
@@ -61,6 +61,32 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
+RUN ruby-install ruby 2.3.1 -- --disable-install-doc && rm -rf /usr/local/src/* && yum clean all
+
+## Build Auth-Api
+RUN mkdir -p /opt/rh/auth-api
```
Should the path be an ENV var, and then you can just use that everywhere?

Fryguy commented on docker-assets/auth-api/Gemfile:
```diff
@@ -0,0 +1,4 @@
+source "https://rubygems.org"
+
+gem "ruby-dbus"
+gem "sinatra"
```
Specific versions?

abellotti committed:
Enhancements to the auth-api service

- Saving PATH in container-environment and leveraging it for running ruby.
- Removing the need for /usr/bin/auth-api-service.sh.
- auth-api.service runs bundle exec ruby directly.
- sinatra version compatibility specified in the Gemfile.
- Defining and using a single HTTPD_AUTH_API_SERVICE_DIRECTORY for the service location.
- Moving auth-api service to /opt/auth-api

carbonin commented on Dockerfile:
```diff
@@ -61,6 +61,32 @@ RUN (cd /lib/systemd/system/sysinit.target.wants && for i in *; do [ $i == syste
+RUN ruby-install ruby 2.3.1 -- --disable-install-doc && rm -rf /usr/local/src/* && yum clean all
+
+## Build Auth-Api
+ENV HTTPD_AUTH_API_SERVICE_DIRECTORY=/opt/rh/auth-api
```
This shouldn't be rh specific. I think that comment got lost in the other file.

Did we come to a consensus about what the path should be? @bdunne @Fryguy

carbonin replied:
We decided `/opt/auth-api`

abellotti replied:
yep, 👍
